openwebwork/webwork2
# PGProblemEditor and CodeMirror

This pull requests adds the option to use a CodeMIrror based code editor in PGProblemEditor2 and PGProblemEditor3 instead of a vanilla text area.  You can read about CodeMirror here https://codemirror.net/ but it is basically a javascript text editor for code.  New features include: 
-   Line numbers. 
-  Tab inserts spaces and doesn't change focus. 
-  Colors (with extra modifications made for PG)

Potential additional features include
-  Autocompletion 
-  Code Folding
-  Bracket Matching
-  Trailing Space Highlighting
-  Lots of other addons listed here:  https://codemirror.net/doc/manual.html#addons

I don't actually write that many problems so I am open to suggestions about what features would be useful.  (I don't want to add them all because I prefer to keep things simple as possible.)  In particular if people have been editing PG problems using more advanced editors like vim or emacs, what features do you appreciate.  

To Test: 
-  Checkout the branch and edit a problem using the editor 2 and editor 3. 
-  Check that the new editor works and that you can go back to the old one by setting
``` 
$options{PGCodeMirror} = 1;
```
-  Look at the coloring on problems and see if it is sane.
-  Edit some problems and see if its an improvement and if any of the existing plugins would be helpful.
